dbms/src/Flash/Pipeline/Schedule/Tasks/RFWaitTask.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <DataStreams/RuntimeFilter.h>
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Schedule/Tasks/Task.h>
+#include <Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h>
+#include <Storages/DeltaMerge/SegmentReadTaskPool.h>
+
+namespace DB
+{
+class RFWaitTask : public Task
+{
+public:
+    RFWaitTask(
+        const String & req_id,
+        PipelineExecutorStatus & exec_status_,
+        const DM::SegmentReadTaskPoolPtr & task_pool_,
+        int max_wait_time_ms,
+        RuntimeFilteList && waiting_rf_list_,
+        RuntimeFilteList && ready_rf_list_)
+        : Task(nullptr, req_id)
+        , exec_status(exec_status_)
+        , task_pool(task_pool_)
+        , max_wait_time_ns(max_wait_time_ms < 0 ? 0 : 1000000UL * max_wait_time_ms)
+        , waiting_rf_list(std::move(waiting_rf_list_))
+        , ready_rf_list(std::move(ready_rf_list_))
+    {
+        exec_status.incActiveRefCount();
+    }
+
+    ~RFWaitTask() override
+    {
+        // In order to ensure that `PipelineExecutorStatus` will not be destructed before `RFWaitTask` is destructed.
+        exec_status.decActiveRefCount();
+    }
+
+    static void filterAndMoveReadyRfs(RuntimeFilteList & waiting_rf_list, RuntimeFilteList & ready_rf_list)
+    {
+        for (auto it = waiting_rf_list.begin(); it != waiting_rf_list.end();)
```

Comment:
It seems that it is just a simple judgment of whether the RF in the current state is ready, so how is the waiting realized?

Reply:
`RFWaitTask` implements the `Task` interface, and `Task::await` will be polled and called in `WaitReactor`.
`RFWaitTask::awaitImpl` maintains a `StopWatch` to know how long it has been "waiting", and decide whether to terminate the "wait".

Reply:
And some comments added.

Reply:
Resolved